### PR TITLE
Add is_empty() for PathBuf (fixes #30259), Overide methods in iterator implementation for EscapeDefault (#24214), Add help note for E0514, Improve internal docs in librustc

### LINKED ISSUE #30259: `PathBuf` does not have `is_empty()`
It is possible to create an empty `PathBuf`:

    let p = PathBuf::new();

— yet `PathBuf` has no `is_empty()` function (nor, apparently, do `Path` or `OsStr` or `OsString`).

This seems inconsistent.

Of course, there are several ways around this (`p == PathBuf::new()`, `p.to_str().map_or(true, |s| s.is_empty())`, wrapping with `Option`), but having a simple way to check this would be nice.

### PR DESCRIPTION
**Pull request by ticki**
See also #30259, #24214 and #30363.

### CODE REVIEW

File: src/compiler-rt [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit b6087e82ba1384c4af3adf2dc68e92316f0d4caf
```
Comment (jonas-schievink) [inline src/compiler-rt]: Looks like a git accident

Comment (Ticki) [inline src/compiler-rt]: I've updated the submodules.

File: src/libstd/path.rs [hunk 2]
```diff
@@ -1859,6 +1860,27 @@ impl Path {
...
+    /// ```
+    /// use std::path::Path;
+    ///
+    /// let path = Path::new("/tmp/foo.rs");
+    ///
+    /// assert!(!path.is_empty());
+    /// ```
+    #[unstable(feature = "path_extras", reason = "recently added", issue = "30259")]
+    pub fn is_empty(&self) -> bool {
+        if let Some(b) = self.inner.to_bytes() {
```
Comment (bluss) [inline src/libstd/path.rs]: I think it's better to use `os_str_as_u8_slice` and check if it is empty. It's not fallible and it's simple.

**Comment by Ticki** [comment]
Oh fuck, I failed and added the naked patch from jackpot51. **Do not merge!**

**Comment by Ticki** [comment]
It's fixed now. I pulled the naked branch on accident (which we use in the Redox rust fork).

**Comment by Ticki** [comment]
I refactored this patch a bit, by adding a `is_empty()` method to `OsStr`, and using it in `Path`, this gives a minor speed-up on Windows.

File: src/libstd/path.rs (hunk 2 above)
Comment (Ticki) [inline src/libstd/path.rs]: @bluss Check out the latest commit.

**Comment by Ticki** [comment]
I'm not sure why travis is failing. It seems to be broken?

**Comment by bluss** [comment]
yep, travis failed to build jemalloc for some reason (so not related to this PR).

Can you put the information in the PR title into the PR message itself? Probably give the PR a shorter title too.

Adding OsStr::is_empty() seems logical, and Path::is_empty looks good too

**Comment by alexcrichton** [comment]
It looks like there's quite a few changes going on here, namely:

1. Lots of submodules updates -- can these all be reverted? It looks like they may be intentional?
2. Specialization of `EscapeDefault` iterator methods -- updated, but no tests have been added. Perhaps this could be separated to a different PR?
3. Various formatting/documentation here and there -- could this be separated to a different PR?
4. Updating a compiler message -- can this also be separated?
5. Addition of new APIs on `Path` and `OsString` -- can this PR be focused on these additions?

This is touching quite a few components which will be easier to review if split apart (as different reviewers are most applicable for the various areas here).

**Comment by Ticki** [comment]
> Lots of submodules updates -- can these all be reverted? It looks like they may be intentional?

Right, I'll downgrade them.

> Specialization of EscapeDefault iterator methods -- updated, but no tests have been added. Perhaps this could be separated to a different PR?

Yeah, will add some tests.

> Various formatting/documentation here and there -- could this be separated to a different PR?

> Updating a compiler message -- can this also be separated?

> Addition of new APIs on Path and OsString -- can this PR be focused on these additions?

Yeah, I'll fix the points above and then seperate them into multiple PRs. Sorry for that.

**Comment by Ticki** [comment]
The submodules are now downgraded.

**Comment by Ticki** [comment]
Oh, I'm good at fucking up with git :(

**Comment by Ticki** [comment]
I'm not sure why travis fails.

> test result: 2272 passed; 0 failed; 52 ignored; 0 measured

**Comment by bluss** [comment]
This is somewhere in the log.

```

failures:

---- path::Path::is_empty_0 stdout ----
	<anon>:8:19: 8:29 error: use of unstable library feature 'os_extras': recently added (see issue #30259)
<anon>:8     assert!(!path.is_empty());
                           ^~~~~~~~~~
<anon>:8:5: 8:31 note: in this expansion of assert! (defined in <std macros>)
<anon>:8:19: 8:29 help: add #![feature(os_extras)] to the crate attributes to enable
error: aborting due to previous error
thread 'path::Path::is_empty_0' panicked at 'Box<Any>', src/libsyntax/errors/mod.rs:269


failures:
    path::Path::is_empty_0
```

**Comment by Ticki** [comment]
Yeah, I realized that I forgot to push the latest commit.

**Comment by nagisa** [comment]
Still seems to be a mess?

**Comment by Ticki** [comment]
Yes. I'll organize this better when I get time.

**Comment by Ticki** [comment]
I've now parted this into multiple PRs. Will close.

**Pull request closed without merging**